Satellite template for OCS (#38)
IBM/ibm-satellite-storage: config-templates/ibm/ocs/README.md

# OCS

Red Hat OpenShift Container Storage is a software-defined storage that is optimised for container environments. It runs as an operator on OpenShift Container Platform to provide highly integrated and simplified persistent storage management for containers.

The user has to provide the input values to the custom resource OcsCluster while creating the satellite configuration to deploy OCS

> ## How to use the local template to deploy OCS on a Satellite Cluster having local storage?

Author: Under ocs we have 2 flavors local and remote for each version. We'll have a common readme as most of the things are common for both the flavors. We will have a remote template also soon.

### Prerequisites

1) In order to deploy OCS, at least one of these local storage options are required:
    - Raw devices (no partitions or formatted filesystems)
    - Raw partitions (no formatted filesystem)
    - PVs available from a storage class in block mode
2) The cluster needs to have a minimum of 3 nodes
3) The OCP version should be compatible with the OCS version you're trying to install
4) The nodes used for the cluster should have a configuration of minimum 16CPUs and 64GB RAM

> ### OCS local: Parameters

Author: Under ocs we have 2 flavors local and remote for each version. We'll have a common readme as most of the things are common for both the flavors. We will have a remote template also soon.

**Description of the template parameters :**

| **Parameter Name** |**Description** |**Datatype** |**Mandatory** |**Default Value** |
|--------------------|----------------|-------------|--------------|------------------|
|ocs-cluster-name | Name of the CRD (OcsCluster)   |string |Yes   |                  |
|mon-device-path      |Devices present on worker to be used for mon pods|csv |Yes |                  |
|osd-device-path      |Devices present on worker to be used for OSD pods  | csv|Yes |                  |

> |num-of-osd        |Number of OSDs required. OCS will create 3x number of OSDs for the value specified here |integer |No |1 |

Author: Num of OSD is related to the storage capacity. Initially, the storage capacity is the disk size of the disk specified at `osd-device-path`. When the user wants to expand the capacity, they can add devices and increase `num-of-osd`.

Example :

```
$ibmcloud sat storage config create --name ocs-config2 --template-name ocs --template-version 4.6_local -p "ocs-cluster-name=testocscluster" -p "osd-device-path=/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part2,/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part3" -p "mon-device-path=/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part1" -p "num-of-osd=2" -p "worker-nodes=169.48.170.83,169.48.170.88,169.48.170.90" -p "ibm-cos-access-key=xxx" -p "ibm-cos-secret-key=yyy"
```

After this, we need to create a new assignment for this configuration :

```
$ ibmcloud sat storage assignment create --name ocs-sub2 --group test-group2 --config ocs-config2
```

> ### Migration (Updating OCS version) :

Reviewer: Instead of `Migration` change it to `How to update OCS version?`

To update the version of OCS, we have to create a new configuration with the template version set to the version of OCS we want to upgrade to. We need to have the same name for the CRD (the parameter `ocs-cluster-name`) and the rest of the parameters should remain the same as the previous configuration, but, we should change the value of `ocs-upgrade` to true.

Example :

```
$ibmcloud sat storage config create --name ocs-config3 --template-name ocs --template-version 4.7_local -p "ocs-cluster-name=testocscluster" -p "osd-device-path=/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part2,/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part3" -p "mon-device-path=/dev/scsi-3600605b00d87b43027b3bc310a64c6c9-part1" -p "num-of-osd=2" -p "worker-nodes=169.48.170.83,169.48.170.88,169.48.170.90" -p "ibm-cos-access-key=xxx" -p "ibm-cos-secret-key=yyy" -p "ocs-upgrade=true"
```

After this, we need to create a new assignment for this configuration :

```
$ ibmcloud sat storage assignment create --name ocs-sub3 --group test-group2 --config ocs-config3
```

This will upgrade OCS to the new version.

> **Note: Please DO NOT delete any of the older configurations and assignments as it may result in DATA LOSS**

Reviewer 2: Let's put this higher up - This should be the first note under the **Upgrading** section

> **Note: Please DO NOT delete any of the older configurations and assignments as it may result in DATA LOSS**

Reviewer 2: ```suggestion
**Important: Do not delete your storage configurations or assignments. Deleting configurations and assignments might result in data loss.**
```

## Troubleshooting

> ### What to do when OCS install fails

Reviewer 2: Do we have any example failure messages? How do they know it fails? Pods stuck in pending, etc...

```suggestion
If your OCS installation fails, complete the following the steps to troubleshoot your deployment.
```

> - Check the describe of storagecluster and cephcluster in the openshift-storage namespace

Reviewer 2: Let's show an output screen from these describe commands. Also, what messages or statuses are they looking for when they run these commands?

```suggestion
1. Describe the `storagecluster` and `cephcluster` in the `openshift-storage` namespace.
```
